core/src/main/java/org/infinispan/topology/ClusterCacheStatus.java

```diff
@@ -899,11 +901,116 @@ public boolean resolveConflictsOnMerge() {
    }
 
    @Override
-   public ConsistentHash calculateConflictHash(ConsistentHash preferredHash, Set<ConsistentHash> distinctHashes) {
+   public ConsistentHash calculateConflictHash(ConsistentHash preferredHash, Set<ConsistentHash> distinctHashes,
+                                               List<Address> actualMembers) {
       // If we are required to resolveConflicts, then we utilise a union of all distinct CHs. This is necessary
       // to ensure that we read the entries associated with all possible read owners before the rebalance occurs
       ConsistentHashFactory chf = getJoinInfo().getConsistentHashFactory();
       ConsistentHash unionHash = distinctHashes.stream().reduce(preferredHash, chf::union);
-      return chf.union(unionHash, chf.rebalance(unionHash));
+      unionHash = chf.union(unionHash, chf.rebalance(unionHash));
+      return chf.updateMembers(unionHash, actualMembers, capacityFactors);
+   }
+
+   @Override
+   public synchronized void queueConflictResolution(CacheTopology conflictTopology, Set<Address> preferredNodes) {
+      if (resolveConflictsOnMerge()) {
+         log.debugf("Queueing conflict resolution for cache %s with members %s", cacheName, conflictTopology.getMembers());
+         queuedConflictResolution = new ConflictData(conflictTopology, preferredNodes);
+         startQueuedConflictResolution();
+      }
+   }
+
+   private synchronized void startQueuedConflictResolution() {
+      if (queuedConflictResolution == null)
+         return;
+
+      if (!conflictFuture.isDone()) {
+         // Check membership, if node removed then we must cancel the CR and start again below
+         // If currentTopology members size is one, cancel CR as it will be re-triggered on future merge
+         conflictFuture.cancel(true);
+      }
+
+      AdvancedCache<?, ?> cache = cacheManager.getCache(cacheName).getAdvancedCache();
+      InternalConflictManager<?, ?> conflictManager = (InternalConflictManager) ConflictManagerFactory.get(cache);
+      conflictFuture = conflictManager.resolveConflicts(queuedConflictResolution.conflictTopology, queuedConflictResolution.preferredNodes);
+      // Will execute using the StateTransferExecutor on completion
+      conflictFuture.thenRun(this::completeConflictResolution);
```

Comment:
I think we also need `completeConflictResolution` to run if CR finishes with a `TimeoutException` (Diego's scenario) or basically anything other than a `SuspectException`.

I was thinking that we may need special handling for `InterruptedException`, too, but then I remembered `CompletableFuture.cancel()` doesn't do anything, so now I'm thinking that `restartConflictResolution()` may need another way to interrupt CR if necessary. TBH I'm not sure interrupting is necessary anyway, as a node leaving would also trigger a `SuspectException` and stop CR.

Reply:
"CompletableFuture.cancel()" still forces the CompletableFuture to complete with a `CancellationException`, which triggers https://github.com/ryanemerson/infinispan/blob/a2d29dff5e799e7215a0d45524f2baae24dc4886/core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java#L251 that then cancels any ongoing segment requests. Or am I missing something?

Reply:
`cancel()` only completes the CF instance you're calling it on. If you want to call the callback added by `resolveConflicts()` with a `CancellationException` you'd have to cancel the initial future, not the result of `future.whenComplete(...)`, but returning the initial future and adding another callback on that would again complicate things by not guaranteeing an execution order.